cranelift/codegen/src/isa/zkasm/lower.isle

```diff
@@ -354,8 +354,8 @@
 
 ;;;;;  Rules for `symbol_value`;;;;;;;;;
 (rule
-   (lower (symbol_value (symbol_value_data name _ offset)))
-   (load_ext_name name offset))
+   (lower (symbol_value (zkasm_base ZkasmBase.Heap)))
+   (imm $I32 0))
```

Comment:
The TODO that was removed from emit.rs mentions having at most one external constant. Hardcoding 0 here means that still at most one external constant is expected?

Reply:
Not quite. The TODO was out of date, as we support 3 types of external constants now: https://github.com/near/wasmtime/blob/76054740a2487ad744ece76d034d13537462bbdf/cranelift/codegen/src/isa/zkasm/lower/isle.rs#L373

But these constants appear in slightly different contexts:
- For `Heap`, we use RegOffset addressing mode
- For `Globals`, we use Global addressing mode
- `Table` is currently not implemented
I looked for a way to handle the `Heap` and `Globals` access similarly, but haven't found it because the generated CLIF code for them is slightly different.

Reply:
Thank you for the context. I'm struggling to understand why there is a hardcoded 0 on the right hand side of this ISLE rule (instead of any other number)?

Reply:
Ah, I see. We use 0 here as the memory region that we use on ZKASM machine starts at offset 0. The offset in global value is only used to identify which memory region we want to access (Heap, Globals, Table). Does this make sense?